src/object/srv_obj.c

```diff
@@ -2022,6 +2022,39 @@ obj_ioc_init_oca(struct obj_io_context *ioc, daos_obj_id_t oid)
 	return 0;
 }
 
+static int
+obj_inflight_io_check(struct ds_cont_child *child, uint32_t opc, uint32_t flags)
+{
+	if (!obj_is_modification_opc(opc))
+		return 0;
+
+	/* If the incoming I/O is during integration, then it needs to wait the
+	 * vos discard to finish, which otherwise might discard these new inflight
+	 * I/O update.
+	 */
+	if ((flags & ORF_REINTEGRATING_IO) &&
+	    (child->sc_pool->spc_pool->sp_need_discard &&
+	     child->sc_pool->spc_discard_done == 0)) {
+		D_ERROR("reintegrating "DF_UUID" retry.\n", DP_UUID(child->sc_pool->spc_uuid));
+		return -DER_UPDATE_AGAIN;
+	}
+
+	/* All I/O during rebuilding, needs to wait for the rebuild fence to
+	 * be generated (see rebuild_prepare_one()), which will create a boundary
+	 * for rebuild, so the data after boundary(epoch) should not be rebuilt,
+	 * which otherwise might be written duplicately, which might cause
+	 * the failure in VOS.
+	 */
+	if ((flags & ORF_REBUILDING_IO) &&
+	    (!child->sc_pool->spc_pool->sp_disable_rebuild &&
+	      child->sc_pool->spc_rebuild_fence == 0)) {
```

Comment:
just confirm, if at the time the rebuild happened to be finished and then set the spc_rebuild_fence = 0, for this case client will refresh pool map and update again? if client don't do the pool map refresh it may dead loop? 
do we need to always set "pm_stale = true" for "-DER_UPDATE_AGAIN" retry case to avoid that? (not very if "obj_auxi->map_ver_reply > obj_auxi->map_ver_req" check can work for all cases).

Reply:
> 

> just confirm, if at the time the rebuild happened to be finished and then set the spc_rebuild_fence = 0, for this case client will refresh pool map and update again? if client don't do the pool map refresh it may dead loop?

If the I/O has the REBUILDING flag, then it means the server will do rebuild anyway, either 

1. rebuild is finished, then client will refresh the pool map, REBUILDING flag will be removed. If the client can not get the refreshed pool map, it will not be able to do update and in endless loop anyway.2. 
2. or rebuild is being started, spc_rebuild_fence is set.

> do we need to always set "pm_stale = true" for "-DER_UPDATE_AGAIN" retry case to avoid that? (not very if "obj_auxi->map_ver_reply > obj_auxi->map_ver_req" check can work for all cases).

Hmm, UPDATE_AGAIN does not mean the pool map needs to refresh, which only means server is busy with this sth, and can not accept update request, vos_discard, waiting for rebuild fence etc.